hw3.1/src/index.ts

```diff
@@ -0,0 +1,237 @@
+  }
+
+  get levels(): Level[] {
+    return this._levels;
+  }
+
+  get name(): string {
+    return this._name;
+  }
+
+  addLevel(level: Level): void {
+    this._levels.push(level);
+  }
+
+  removeLevel (level: Level): void {
+    this._levels = this.levels.filter(item => item != level)
+  }
+}
+ 
+class Level {
+  // implement getters for fields and 'add/remove group' methods
+
+  private _groups: Group[] = [];
+  private _name: string;
+  private _description: string;
+
+  constructor(name: string, description: string) {
+    this._name = name;
+    this._description = description;
+  }
+
+  get name(): string {
+    return this._name;
+  }
+  
+  get groups(): Group[] {
+    return this._groups;
+  }
+
+  get description(): string {
+    return this._description;
+  }
+
+  addGroup(group: Group): void {
+    this._groups.push(group);
+  }
+
+  removeGroup (group: Group): void {
+    this._groups = this.groups.filter(item => item != group)
+  }
+}
+ 
+class Group {
+  // implement getters for fields and 'add/remove student' and 'set status' methods
+
+  private _area: string;
+  private _status: string;
+  private _level: string;
+  private _students: Student [] = []; // Modify the array so that it has a valid toSorted method*
+
+  constructor(area: string, level: string, status: string) {
+    this._area = area; 
+    this._level = level;
+    this._status = status;
+  }
+
+  get area(): string {
+    return this._area;
+  }
+
+  get status(): string {
+    return this._status;
+  }
+
+  get level(): string {
+    return this._level;
+  }
+
+  set status(value: string) {
+    this._status = value;
+  }
+
+  addStudent(student: Student): void {
+    this._students.push(student);
+  }
+
+  removeStudent(student: Student): void {
+    this._students = this._students.filter(s => s !== student);
+  }
+
+  showPerformance() {
+    const sortedStudents = this._students.sort((a, b) => b.getPerformanceRating() - a.getPerformanceRating());
+    return sortedStudents;
+  }
+}
+ 
+class Student {
+  // implement 'set grade' and 'set visit' methods
+
+  private _firstName: string;
+  private _lastName: string;
+  private _birthYear: number;
+  private _grades: { [workName: string]: number } = {}; // workName: mark
+  private _visits: { [lesson: string]: boolean } = {}; // lesson: present
+
+  constructor(firstName: string, lastName: string, birthYear: number) {
+    this._firstName = firstName;
+    this._lastName = lastName;
+    this._birthYear = birthYear;
+  }
+
+  get fullName(): string {
+    return `${this._lastName} ${this._firstName}`;
+  }
+
+  set fullName(value: string) {
+    [this._lastName, this._firstName] = value.split(' ');
+  }
+
+  get age(): number {
+    return new Date().getFullYear() - this._birthYear;
+  }
+
+  setGrade (workName: string, mark: number) {
+    this._grades[workName] = mark;
+  }
+
+  setVisit (lesson: string, present: boolean) {
+    this._visits[lesson] = present;
+  }
+
+  getPerformanceRating() {
```

Comment:
Don't forget to include explicit type annotations for the return function value.